opds serializable  classes

Prashant-Android opened the pull request.

mikedawson commented on shared/src/commonMain/kotlin/world/respect/model/OpdsSerialization.kt:
This class should not exist - it provides no value or benefit vs. using kotlinx's own Json

mikedawson commented on shared/src/commonTest/kotlin/world/respect/OpdsFileParsingTest.kt:
```diff
@@ -0,0 +1,49 @@
+    @Test
+    fun testParseSampleFiles() {
+        // Get the paths to your sample JSON files
+        val sampleCatalogPath = "/home/prashant/AndroidStudioProjects/Respect/composeApp/src/desktopMain/resources/sample-catalog.json"
```
Are. You. Kidding. Me?

mikedawson commented on shared/src/jvmMain/kotlin/world/respect/OpdsCliApp.kt:
```diff
@@ -0,0 +1,215 @@
+     *
+     * @param filePath The path to the OPDS file
+     */
+    private fun validateOpdsFile(filePath: String) {
```
Using a filePath here is wrong: OPDS feeds are typically published on http. This should accept a URI. The validator will have to follow links in the OPDS.

mikedawson commented on shared/src/jvmMain/kotlin/world/respect/OpdsCliApp.kt:
```diff
@@ -0,0 +1,215 @@
+     * @param filePath The path to the OPDS file
+     */
+    private fun validateOpdsFile(filePath: String) {
+        try {
```
Validation logic should not be here (in command line interface/CLI) - it should be in the UseCase . The use case could be used by CLI, could be used by the RESPECT app itself, etc.

mikedawson commented on shared/src/commonMain/kotlin/world/respect/domain/opds/validator/OpdsValidator.kt:
```diff
@@ -0,0 +1,168 @@
+        }
+
+        // Check for acquisition links
+        val acquisitionRelations = listOf(
```
These should be constants defined in a sensible place coding style).

mikedawson commented on shared/src/commonMain/kotlin/world/respect/domain/opds/model/OpdsFeedMetadata.kt:
```diff
@@ -0,0 +1,27 @@
+ */
+@Serializable
+data class OpdsFeedMetadata(
+    val title: String,                     // Required field (title of the catalog or collection)
```
@Prashant-Android This is missing the @type property as required by https://drafts.opds.io/schema/feed-metadata.schema.json . You can use Kotlinx serialization annotations if the field name in Json is not a valid variable name in Kotlin.

mikedawson replied:
Also missing identifier.

mikedawson replied:
Also missing subtitle property

mikedawson commented on shared/src/commonMain/kotlin/world/respect/domain/opds/validator/OpdsValidator.kt:
```diff
@@ -0,0 +1,601 @@
+class OpdsValidator {
+
+    // List to store warnings that don't cause validation failure
+    private val warnings = mutableListOf<String>()
```
Clearly and obviously wrong: why is this a member variable on the class? The list of warnings needs built up per validation run.

mikedawson commented on shared/src/commonMain/kotlin/world/respect/domain/opds/validator/OpdsValidator.kt:
```diff
@@ -0,0 +1,601 @@
+    // Json instance for serialization and deserialization
+    private val json = Json {
+        ignoreUnknownKeys = true  // Ignores unknown keys in JSON for compatibility
+        isLenient = true          // Allows lenient parsing of the data
```
Why would a validator be lenient? This is wrong. The Json instance should be a constructor parameter.

Prashant-Android commented:
@mikedawson 

here is the updates: 
1. Added missing properties to OpdsFeedMetadata: identifier, @type (with @SerialName), and subtitle.
2. Fixed warnings handling by removing the class-level list and using a per-validation warnings collection.
3. Made the Json instance a constructor parameter.

cmd line : 
./gradlew runOpdsValidator --args="https://example.com/test.json"

mikedawson commented on shared/src/commonMain/kotlin/world/respect/domain/opds/model/OpdsLink.kt:
```diff
@@ -0,0 +1,26 @@
+ * For reference, see the schema: https://readium.org/webpub-manifest/schema/link.schema.json
+ */
+@Serializable
+data class OpdsLink(
```
This is also missing properties from the spec as per https://readium.org/webpub-manifest/schema/link.schema.json

e.g. size, bitrate, language, alternate

As mentioned: please ensure that **all** properties are included 100% as per the spec.

mikedawson commented on shared/src/commonMain/kotlin/world/respect/domain/opds/model/OpdsFeedMetadata.kt:
```diff
@@ -0,0 +1,35 @@
+ * Represents metadata for an OPDS catalog or collection.
+ * This includes details like title, number of items, items per page, etc.
+ *
+ * For reference, see the schema: https://github.com/readium/webpub-manifest/blob/master/schema/metadata.schema.json
```
Wrong reference link for schema: for feed should be https://drafts.opds.io/schema/feed-metadata.schema.json

mikedawson commented on shared/src/commonMain/kotlin/world/respect/domain/opds/model/OpdsPublicationMetadata.kt:
```diff
@@ -0,0 +1,31 @@
+
+/**
+ * Represents metadata for a Publication.
+ * For reference, see the schema: https://drafts.opds.io/opds-2.0#52-metadata
```
The schema is actually: https://readium.org/webpub-manifest/schema/metadata.schema.json

Prashant-Android replied:
done

mikedawson commented on shared/src/commonMain/kotlin/world/respect/domain/opds/model/OpdsPublicationMetadata.kt:
```diff
@@ -0,0 +1,31 @@
+    val language: String? = null,           // Optional field (language of the publication)
+    val modified: String? = null,           // Optional field (last modification date)
+    val description: String? = null,        // Optional field (description of the publication)
+    val belongsTo: OpdsBelongsTo? = null    // Optional field (indicates series or collection)
```
Missing: readingDirection, numberOfPages, subject

mikedawson commented:
@Prashant-Android  - to actually do the validation - I think it would be much better to use an existing JSON schema validation library as per https://json-schema.org/tools?query=&sortBy=name&sortOrder=ascending&groupBy=toolingTypes&licenses=&languages=&drafts=&toolingTypes=&environments=&showObsolete=false e.g. https://github.com/harrel56/json-schema or https://github.com/OptimumCode/json-schema-validator 

We will still need to have Kotlin data classes for the app itself: but for the validation, this only needs to be done server/command line side. So using an existing Kotlin or Java based JSON schema validator would do this better.

Prashant-Android commented:
> @Prashant-Android - to actually do the validation - I think it would be much better to use an existing JSON schema validation library as per https://json-schema.org/tools?query=&sortBy=name&sortOrder=ascending&groupBy=toolingTypes&licenses=&languages=&drafts=&toolingTypes=&environments=&showObsolete=false e.g. https://github.com/harrel56/json-schema or https://github.com/OptimumCode/json-schema-validator
> 
> We will still need to have Kotlin data classes for the app itself: but for the validation, this only needs to be done server/command line side. So using an existing Kotlin or Java based JSON schema validator would do this **much** better.

The json-schema-validator will also validate recursive links.
